component/grant-type/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/TokenExchangeGrantValidator.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License
+ */
+
+package org.wso2.carbon.identity.oauth2.grant.token.exchange;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.oltu.oauth2.common.OAuth;
+import org.apache.oltu.oauth2.common.error.OAuthError;
+import org.apache.oltu.oauth2.common.exception.OAuthProblemException;
+import org.apache.oltu.oauth2.common.validators.AbstractValidator;
+
+import javax.servlet.http.HttpServletRequest;
+
+import static org.wso2.carbon.identity.oauth.common.OAuthCommonUtil.validateContentTypes;
+
+/**
+ * Grant validator for Token Exchange Grant Type.
+ * A token exchange grant request should have the required parameters -
+ * grant_type, subject_token and subject_token_type.
+ */
+public class TokenExchangeGrantValidator extends AbstractValidator<HttpServletRequest> {
+
+    private static final Log log = LogFactory.getLog(TokenExchangeGrantValidator.class);
+
+    public TokenExchangeGrantValidator() {
+        requiredParams.add(OAuth.OAUTH_GRANT_TYPE);
+        requiredParams.add(TokenExchangeConstants.SUBJECT_TOKEN);
+        requiredParams.add(TokenExchangeConstants.SUBJECT_TOKEN_TYPE);
+    }
+
+    /**
+     * Validate the content-type of the request
+     *
+     * @param request - HttpServletRequest
+     * @throws OAuthProblemException Exception when validating the grant type
+     */
+    @Override
+    public void validateContentType(HttpServletRequest request) throws OAuthProblemException {
+
+        validateContentTypes(request);
+    }
+
+    /**
+     * Validate the required request parameters - subject_token_type, subject_token and grant_type in payload
+     *
+     * @param request - HttpServletRequest
+     * @throws OAuthProblemException Exception when validating the required parameters
+     */
+    @Override
+    public void validateRequiredParameters(HttpServletRequest request) throws OAuthProblemException {
+        super.validateRequiredParameters(request);
+        String subjectTokenType = request.getParameter(TokenExchangeConstants.SUBJECT_TOKEN_TYPE);
+        if(!StringUtils.equals(TokenExchangeConstants.JWT_TOKEN_TYPE, subjectTokenType)) {
```

Comment:
What if someone sends a JWT token with the token type set to `urn:ietf:params:oauth:token-type:access_token`

Reply:
Fixed it. Changed the impl to check the token type and if it is urn:ietf:params:oauth:token-type:access_token then it will check if it's JWT. If it is not JWT, then error will be thrown